ghcide/src/Development/IDE/GHC/Compat/Driver.hs

```diff
@@ -0,0 +1,144 @@
+-- ============================================================================
+-- DO NOT EDIT
+-- This module copies parts of the driver code in GHC.Driver.Main to provide
+-- `hscTypecheckRenameWithDiagnostics`.
+-- Issue to add this function: https://gitlab.haskell.org/ghc/ghc/-/issues/24996
+-- MR to add this function: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12891
+-- ============================================================================
+
+{-# LANGUAGE CPP #-}
+
+module Development.IDE.GHC.Compat.Driver
+    ( hscTypecheckRenameWithDiagnostics
+    ) where
+
+import           Control.Monad
+import           GHC.Core
+import           GHC.Data.FastString
+import           GHC.Data.Maybe
+import           GHC.Driver.Env
+import           GHC.Driver.Errors.Types
+import           GHC.Driver.Main
+import           GHC.Driver.Session
+import           GHC.Hs
+import           GHC.Hs.Dump
+import           GHC.Iface.Ext.Ast               (mkHieFile)
+import           GHC.Iface.Ext.Binary            (hie_file_result, readHieFile,
+                                                  writeHieFile)
+import           GHC.Iface.Ext.Debug             (diffFile, validateScopes)
+import           GHC.Iface.Ext.Types             (getAsts, hie_asts, hie_module)
+import           GHC.Tc.Module
+import           GHC.Tc.Utils.Monad
+import           GHC.Types.SourceFile
+import           GHC.Types.SrcLoc
+import           GHC.Unit
+import           GHC.Unit.Module.ModDetails
+import           GHC.Unit.Module.ModIface
+import           GHC.Unit.Module.ModSummary
+import           GHC.Utils.Error
+import           GHC.Utils.Logger
+import           GHC.Utils.Outputable
+import           GHC.Utils.Panic.Plain
+
+#if !MIN_VERSION_ghc(9,6,1)
+import           Development.IDE.GHC.Compat.Core (hscTypecheckRename)
+import           GHC.Utils.Error                 (emptyMessages)
+#endif
+
+hscTypecheckRenameWithDiagnostics :: HscEnv -> ModSummary -> HsParsedModule
+                   -> IO ((TcGblEnv, RenamedStuff), Messages GhcMessage)
+hscTypecheckRenameWithDiagnostics hsc_env mod_summary rdr_module =
+#if MIN_VERSION_ghc(9,6,1)
+    runHsc' hsc_env $ hsc_typecheck True mod_summary (Just rdr_module)
+#else
+    (,emptyMessages) <$> hscTypecheckRename hsc_env mod_summary rdr_module
```

Comment:
Through the magic of inserting `false &&` in these CPP blocks, I've figured out this is the problematic one. But it is still mysterious to me that this causes errors; it seems very benign.

Reply:
@fendor I'm getting stuck debugging this. Every caller of this typechecking code path seems to complete fine all the way up to the shake rules. However, I don't see where the resulting LSP message is supposed to be produced.

Reply:
`emptyMessages` looks suspicious to me, maybe we are not propagating diagnostics occurring in the renamer phase, not producing code actions for all messages originating from that phase?

Reply:
My thoughts were that it returns no messages because diagnostic numbers were only introduced in 9.6, but maybe that is indeed not the whole story. I wonder if 9.4 ever worked with the original PR.

Reply:
The messages are presumably the diagnostics, not the error codes themselves, right? 

I can try to take a closer look some time this week, but I am not terribly familiar with how it used to work.

Reply:
Yes, that does sound fishy. I'll take another look with that in mind.

Reply:
I think it was just a dummy implementation. Now I've changed it to do the same thing as we do for 9.6 and it seems to work locally for me.